pilot/pkg/status/resourcelock.go

```diff
@@ -14,47 +14,168 @@
 
 package status
 
-import "sync"
+import (
+	"context"
+	"sync"
+	"time"
 
-type ResourceLock struct {
-	masterLock sync.RWMutex
-	listing    map[interface{}]*sync.Mutex
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"istio.io/pkg/log"
+)
+
+// Task to be performed.
+type Task func(entry *cacheEntry) error
+
+// Worker queue implements an expandable goroutine pool which executes at most one concurrent routine per target
+// resource.  Multiple calls to Push() will not schedule multiple executions per target resource, but will ensure that
+// the single execution uses the latest value.
+type WorkerQueue interface {
+	// Push a task.
+	Push(target Resource, progress Progress)
+	// Run the loop until a signal on the context
+	Run(ctx context.Context)
+	// Delete a task
+	Delete(target lockResource)
 }
 
-func (r *ResourceLock) Lock(i interface{}) {
-	r.retrieveMutex(i).Lock()
+type queueImpl struct {
+	delay   time.Duration
+	tasks   []lockResource
+	cond    *sync.Cond
+	closing bool
+	work    Task
+
+	cacheLock   sync.Mutex
+	cache       map[lockResource]*cacheEntry
+	workerCount int
+	maxWorkers  int
 }
 
-func (r *ResourceLock) Unlock(i interface{}) {
-	r.retrieveMutex(i).Unlock()
+// NewQueue instantiates a queue with a processing function
+func NewQueue(errorDelay time.Duration, work Task, maxWorkers int) WorkerQueue {
+	return &queueImpl{
+		delay:       errorDelay,
+		tasks:       make([]lockResource, 0),
+		closing:     false,
+		cond:        sync.NewCond(&sync.Mutex{}),
+		work:        work,
+		workerCount: 0,
+		maxWorkers:  maxWorkers,
+		cache:       make(map[lockResource]*cacheEntry),
+	}
 }
 
-// returns value indicating if init was necessary
-func (r *ResourceLock) init() bool {
-	if r.listing == nil {
-		r.masterLock.Lock()
-		defer r.masterLock.Unlock()
-		// double check, per pattern
-		if r.listing == nil {
-			r.listing = make(map[interface{}]*sync.Mutex)
-		}
-		return true
+func (q *queueImpl) Push(target Resource, progress Progress) {
+	q.cacheLock.Lock()
+	defer q.cacheLock.Unlock()
+	key := convert(target)
+	_, inqueue := q.cache[key]
+	q.cache[key] = &cacheEntry{
+		cacheVal:      &target,
+		cacheProgress: &progress,
+	}
+	if !inqueue {
+		q.enqueue(key)
 	}
-	return false
+	q.maybeAddWorker()
 }
 
-func (r *ResourceLock) retrieveMutex(i interface{}) *sync.Mutex {
-	if !r.init() {
-		r.masterLock.RLock()
-		if result, ok := r.listing[i]; ok {
-			r.masterLock.RUnlock()
-			return result
+func (q *queueImpl) Delete(target lockResource) {
+	q.cacheLock.Lock()
+	defer q.cacheLock.Unlock()
+	delete(q.cache, target)
+}
+
+func (q *queueImpl) enqueue(item lockResource) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	if !q.closing {
+		q.tasks = append(q.tasks, item)
+	}
+	q.cond.Signal()
+}
+
+func (q *queueImpl) maybeAddWorker() {
```

Comment:
nit: a few comments here could clear this up. I think the idea is pretty simple: if we aren't at max workers, start a new one. The worker will exit if there is no work, unless its the last one.

But the code is fairly complex (not unnecessarily, it just is dealing with a fairly complex problem) so it takes a couple minutes to grok this fully